internal/lsp/config/watcher.go

```diff
@@ -0,0 +1,136 @@
+package config
+
+import (
+	"context"
+	"fmt"
+	"io"
+
+	"github.com/fsnotify/fsnotify"
+)
+
+type Watcher struct {
+	Reload chan string
+	Drop   chan struct{}
+
+	path        string
+	pathUpdates chan string
+
+	fsWatcher *fsnotify.Watcher
+
+	errorWriter io.Writer
+}
+
+type WatcherOpts struct {
+	ErrorWriter io.Writer
+	Path        string
+}
+
+func NewWatcher(opts *WatcherOpts) *Watcher {
+	w := &Watcher{
+		Reload:      make(chan string, 1),
+		Drop:        make(chan struct{}, 1),
+		pathUpdates: make(chan string, 1),
+	}
+
+	if opts != nil {
+		w.errorWriter = opts.ErrorWriter
+		w.path = opts.Path
+	}
+
+	return w
+}
+
+func (w *Watcher) Start(ctx context.Context) error {
+	var err error
+
+	err = w.Stop()
+	if err != nil {
+		return fmt.Errorf("failed to stop existing watcher: %w", err)
+	}
+
+	w.fsWatcher, err = fsnotify.NewWatcher()
+	if err != nil {
+		return fmt.Errorf("failed to create fsnotify watcher: %w", err)
+	}
+
+	go func() {
+		w.loop(ctx)
+	}()
+
+	return nil
+}
+
+func (w *Watcher) loop(ctx context.Context) {
+	var err error
```

Comment:
[style] Hmm there seems to be a pattern here.. there are a few situations where declaring this upfront is required, like
```go
foo := "y"
if x {
    foo, err = f() // := would introduce a new `foo` in this branch's scope, not what we want
}
```
but outside of those, I think I'd rather go with the `:=`..

Reply:
Thanks, that's fair. I have fixed in b4efd89